interfaces/apparmor/backend.go

```diff
@@ -426,6 +436,59 @@ func (b *Backend) Setup(snapInfo *snap.Info, opts interfaces.ConfinementOptions,
 	return errUnload
 }
 
+func (b *Backend) SetupMany(snaps []*snap.Info, confinement func(snapName string) interfaces.ConfinementOptions, repo *interfaces.Repository, tm timings.Measurer) []error {
+	var allChangedPaths, allUnchangedPaths, allRemovedPaths []string
+	var fallback bool
+	for _, snapInfo := range snaps {
+		opts := confinement(snapInfo.InstanceName())
+		changedPaths, removedPaths, unchangedPaths, err := b.prepareProfiles(snapInfo, opts, repo)
+		if err != nil {
+			fallback = true
+			break
+		}
+		allChangedPaths = append(allChangedPaths, changedPaths...)
+		allUnchangedPaths = append(allUnchangedPaths, unchangedPaths...)
+		allRemovedPaths = append(allRemovedPaths, removedPaths...)
+	}
+
+	if !fallback {
+		var errReloadChanged error
+		timings.Run(tm, "load-profiles-many[changed]", fmt.Sprintf("load changed security profiles of %d snaps", len(snaps)), func(nesttm timings.Measurer) {
+			errReloadChanged = loadProfiles(allChangedPaths, dirs.AppArmorCacheDir, skipReadCache|conserveCPU)
+		})
+
+		var errReloadOther error
+		timings.Run(tm, "load-profiles-many[unchanged]", fmt.Sprintf("load unchanged security profiles %d snaps", len(snaps)), func(nesttm timings.Measurer) {
+			errReloadOther = loadProfiles(allUnchangedPaths, dirs.AppArmorCacheDir, conserveCPU)
+		})
+		errUnload := unloadProfiles(allRemovedPaths, dirs.AppArmorCacheDir)
+		if errReloadChanged != nil {
+			logger.Noticef("failed to reload changed profiles: %s", errReloadChanged)
+			fallback = true
+		}
+		if errReloadOther != nil {
+			logger.Noticef("failed to reload unchanged profiles: %s", errReloadOther)
+			fallback = true
+		}
+		if errUnload != nil {
+			logger.Noticef("failed to unload profiles: %s", errUnload)
+			fallback = true
+		}
```

Comment:
Considering my comment regarding unload only unlinking, errUnload doesn't strictly have to trigger the fallback code. However, we want to get to the point of unloading once snap processes are gone, so doing something else for this case would be prematurely optimizing. Ie, leaving it here is good. :)

Reply:
Indeed, I'd keep it that way for consistency.